### PR TITLE
GraphQL TryOut Console for Developer Portal

### PR DESCRIPTION
**Pull request by rasika1995**
Screenshot from 2020-02-29 18-51-20

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLConsole.jsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
+import React, { useState, useContext } from 'react';
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLConsole.jsx]: Add license header.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/.eslintignore [hunk 2]
```diff
@@ -9,6 +9,7 @@ source/src/app/data/*
 /source/src/app/components/Apis/Details/Comments/*
 /source/src/app/components/Apis/Details/Credentials/*
 /source/src/app/components/Apis/Details/Documents/*
+/source/src/app/components/Apis/Details/GraphQLConsole/*
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/.eslintignore]: Shall we fix any eslit issues found in the components?  We also need to enable other components and fix them as well.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLUI.jsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
+import React, {
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLUI.jsx]: License header missing.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLUI.jsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import TextField from '@material-ui/core/TextField';
+import { FormattedMessage } from 'react-intl';
+import GraphiQLExplorer from 'graphiql-explorer';
+import Api from 'AppData/api';
+import Box from '@material-ui/core/Box';
+import { ApiContext } from '../ApiContext';
+import Progress from '../../../Shared/Progress';
+
+
+const { buildSchema } = require('graphql');
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLUI.jsx]: Remove unnecessary empty lines in this file.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx [hunk 5]
```diff
@@ -0,0 +1,392 @@
+import React, { useState, useEffect } from 'react';
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx]: Missing license header

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+            .then(() => {
+                if (user != null) {
+                    return apiClient.getSubscriptions(apiID);
+                } else {
+                    return null;
+                }
+            })
+            .then((subscriptionsResponse) => {
+                if (subscriptionsResponse != null) { //
+                    const subs = subscriptionsResponse.obj.list.filter((item) => item.status === 'UNBLOCKED' || item.status === 'PROD_ONLY_BLOCKED');
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx]: Exceeds 120 character limit.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+                        const sApplication = subs[0].applicationId;
+                        setSelectedApplication(sApplication);
+                        const promiseApp = Application.get(sApplication);
+                        promiseApp
+                            .then((application) => {
+                                return application.getKeys();
+                            })
+                            .then((appKeys) => {
+                                if (appKeys.get('SANDBOX')) {
+                                    const sKType1 = 'SANDBOX';
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx]: I think skType1 and skType2 are redundant declarations. We can remove them or, define as constants.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+                                    setSelectedKeyType(sKType1);
+                                    ({ accessToken } = appKeys.get('SANDBOX').token);
+                                } else if (appKeys.get('PRODUCTION')) {
+                                    const sKType2 = 'PRODUCTION';
+                                    setSelectedKeyType(sKType2);
+                                    ({ accessToken } = appKeys.get('PRODUCTION').token);
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx]: Are we modifying the props here?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+
+
+    /**
+     *
+     * @param {React.SyntheticEvent} event
+     */
+    const handleChanges = (event) => {
+        if (event.target.name === 'selectedApplication') {
+            const promiseApp = Application.get(event.target.value);
+            let accessToken;
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx]: The same variable is passed through props as well. Use a different name.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+            let keyType;
+
+            setSelectedApplication(event.target.value);
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx]: Use object destruction method. 
 `const { value } = event.target;`

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+
+            if (subscriptions != null && subscriptions.find((sub) => sub.applicationId
+                === selectedApplication).status === 'PROD_ONLY_BLOCKED') {
+                const sKType1 = 'SANDBOX';
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx]: Here as well... Redundant declaration.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+                    <Grid item md={6}>
+                        <Paper className={classes.userNotificationPaper}>
+                            <Typography variant='h5' component='h3'>
+                                <Icon>warning</Icon>
+                                {' '}
+                                <FormattedMessage id='notice' defaultMessage='Notice' />
+                            </Typography>
+                            <Typography component='p'>
+                                <FormattedMessage
+                                    id='api.console.require.access.token'
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx]: Use proper id for FormattedMessages.

`Apis.Details.GraphQLConsole.GraphQLAuthentication.api.console.require.access.token`

Comment (menakaj) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx]: Fix in the other places as well.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/index.jsx [hunk 13]
```diff
@@ -52,6 +53,16 @@ const LoadableSwitch = withRouter((props) => {
     const path = '/apis/';
     const { advertised } = api.advertiseInfo;
     const redirectURL = path + apiUuid + '/overview';
+
+
+    let tryoutRoute;
+    if (api.type === 'GRAPHQL') {
+        tryoutRoute = <Route path='/apis/:apiUuid/test' component={GraphQLConsole} />
+    }
+    else {
```
Comment (menakaj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/index.jsx]: Keep } else { in a single line.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLUI.jsx [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+                        </Box>
+                        <Box display='flex' height='800px' flexGrow={1}>
+                            <GraphiQL
+                                ref={graphiqlEl}
+                                fetcher={(graphQLFetcher)}
+                                schema={schema}
+                                query={query}
+                                // variables={variables}
+                                onEditQuery={setQuery}
+                                // onEditVariables={onEditVariables}
```
Comment (pahans) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLUI.jsx]: remove commented code

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLUI.jsx [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+                    <Box display='flex'>
+                        <Box display='flex'>
+                            <GraphiQLExplorer
+                                schema={schema}
+                                query={query}
+                                onEdit={setQuery}
+                                explorerIsOpen={isExplorerOpen}
+                                onToggleExplorer={handleToggleExplorer}
+                            // onRunOperation={operationName =>
+                            //     this._graphiql.handleRunQuery(operationName)
```
Comment (pahans) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLUI.jsx]: Remove commented code

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLConsole.jsx [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+                    setAccessTocken={setAccessTocken}
+                    authorizationHeader={authorizationHeader}
+                    securitySchemeType={securitySchemeType}
+                    setSecuritySchemeType={setSecuritySchemeType}
+                    prefix={prefix}
+                    isApiKeyEnabled={isApiKeyEnabled}
+                    selectedEnvironment={selectedEnvironment}
+                    setSelectedEnvironment={setSelectedEnvironment}
+                    environments={environments}
+                    setURLss={setURLss}
```
Comment (pahans) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLConsole.jsx]: `setURLss` or `setURLs`?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLConsole.jsx [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+import GraphQLUI from './GraphQLUI';
+import { ApiContext } from '../ApiContext';
+import Progress from '../../../Shared/Progress';
+
+
+const useStyles = makeStyles((theme) => ({
+    paper: {
+        margin: theme.spacing(1),
+        padding: theme.spacing(1),
+        // height: theme.spacing(120),
```
Comment (pahans) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLConsole.jsx]: Remove commented code

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx [hunk 18]
```diff
@@ -0,0 +1,392 @@
...
+    /**
+     * Set the security scheme type for selected security scheme type
+     * @param {*} event
+     */
+    const handlesecuritySchemeType = (event) => {
+        setSecuritySchemeType(event.target.value);
+    };
+
+
+    return (
```
Comment (pahans) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLAuthentication.jsx]: render function is too lengthy and harder to read IMO. suggestion: break this into smaller functions.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLConsole.jsx [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+    const { api } = useContext(ApiContext);
+    const environmentObject = api.endpointURLs;
+    const [URLss, setURLss] = useState(environmentObject[0].URLs);
+    const [accessToken, setAccessTocken] = useState('');
+    const [securitySchemeType, setSecuritySchemeType] = useState('OAUTH');
+    const environments = api.endpointURLs.map((endpoint) => { return endpoint.environmentName; });
+    const [selectedEnvironment, setSelectedEnvironment] = useState(environments[0]);
+    const [notFound, setFound] = useState(false);
+
+
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLConsole.jsx]: Shall we remove this extra line?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLUI.jsx [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+        const apiClient = new Api();
+        const promiseGraphQL = apiClient.getGraphQLSchemaByAPIId(apiID);
+        promiseGraphQL
+            .then((res) => {
+                const graphqlSchemaObj = buildSchema(res.data);
+                setSchema(graphqlSchemaObj);
+            });
+    }, []);
+
+
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/GraphQLUI.jsx]: remove extra line

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/explorer.css [hunk 21]
```diff
@@ -0,0 +1,6 @@
+.graphiql-explorer-root {
+    max-height: 750px;
+    overflow-y: scroll;
+}
+
+
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/GraphQLConsole/explorer.css]: remove extra line

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/index.jsx [hunk 22]
```diff
@@ -52,6 +53,16 @@ const LoadableSwitch = withRouter((props) => {
     const path = '/apis/';
     const { advertised } = api.advertiseInfo;
     const redirectURL = path + apiUuid + '/overview';
+
+
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/index.jsx]: remove extra line

**Pull request merged**